Source/Services.Clients/ReverseCallClient.cs

```diff
@@ -0,0 +1,221 @@
+            IExecutionContextManager executionContextManager,
+            ILogger logger)
+        {
+            _establishConnection = establishConnection;
+            _setConnectArguments = setConnectArguments;
+            _getConnectResponse = getConnectResponse;
+            _getMessageRequest = getMessageRequest;
+            _setMessageResponse = setMessageResponse;
+            _setArgumentsContext = setArgumentsContext;
+            _getRequestContext = getRequestContext;
+            _setResponseContext = setResponseContext;
+            _executionContextManager = executionContextManager;
+            _logger = logger;
+        }
+
+        /// <inheritdoc/>
+        public TConnectResponse ConnectResponse { get; private set; }
+
+        /// <inheritdoc/>
+        public async Task<bool> Connect(TConnectArguments connectArguments, CancellationToken cancellationToken)
+        {
+            if (_alreadyConnected) throw new ReverseCallClientAlreadyCalledConnect();
+            lock (_connectLock)
+            {
+                if (_alreadyConnected) throw new ReverseCallClientAlreadyCalledConnect();
+                _alreadyConnected = true;
+            }
+
+            var streamingCall = _establishConnection();
+            _clientToServer = streamingCall.RequestStream;
+            _serverToClient = streamingCall.ResponseStream;
+            var callContext = new ReverseCallArgumentsContext
+                {
+                    ExecutionContext = _executionContextManager.Current.ToProtobuf()
+                };
+            _setArgumentsContext(connectArguments, callContext);
+            var message = new TClientMessage();
+            _setConnectArguments(message, connectArguments);
+
+            await _clientToServer.WriteAsync(message).ConfigureAwait(false);
+            if (await _serverToClient.MoveNext(cancellationToken).ConfigureAwait(false))
+            {
+                var response = _getConnectResponse(_serverToClient.Current);
+                if (response != null)
+                {
+                    _logger.Trace("Received connect response");
+                    ConnectResponse = response;
+                    return true;
+                }
+                else
+                {
+                    _logger.Warning("Did not receive connect response. Server message did not contain the connect response");
+                }
+            }
+            else
+            {
+                _logger.Warning("Did not receive connect response. Server stream was empty");
+            }
+
+            await _clientToServer.CompleteAsync().ConfigureAwait(false);
+            return false;
+        }
+
+        /// <inheritdoc/>
+        public async Task Handle(Func<TRequest, CancellationToken, Task<TResponse>> callback, CancellationToken cancellationToken)
+        {
+            if (ConnectResponse == null) throw new ReverseCallClientNotConnected();
```

Comment:
In stead of checking for `null` here, we should set a bool property in the `Connect` method above that indicates whether connection has been established. So you can do that just before the `return true` statement e.g.

Reply:
fixed by https://github.com/dolittle-fundamentals/DotNET.Fundamentals/pull/310/commits/43403fe1a3e9cbaf47d70b3e41f57d7ac5f3da9e